### PR TITLE
FIP-0094: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM

### PR DESCRIPTION
**Pull request by snissn**
The `MCOPY` opcode provides a more efficient mechanism for memory copying within smart contracts. This proposal suggests integrating `MCOPY` into the FEVM, ensuring compatibility with Solidity versions v0.8.25 and later, which rely on this opcode for certain operations.

 https://github.com/filecoin-project/FIPs/discussions/1025

### CODE REVIEW

File: FIPS/fip-mcopy-opcode-fevm.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+fip: "<to be assigned>"
+title: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM
+author: Michael Seiler (@snissn), Raúl Kripalani (@raulk), Steven Allen (@stebalien)
+discussions-to: https://github.com/filecoin-project/FIPs/discussions/1025
+status: Draft
+type: Technical
+category: Core
+created: 2024-08-22
+spec-sections:
+  - MCOPY
```
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:11]:
```suggestion
```

I think this section is for the link to spec.filecoin.io - however, i dont think we are updating the spec, especailly for fvm atm. So maybe remove?

File: FIPS/fip-mcopy-opcode-fevm.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+Similar to other memory manipulation operations, `MCOPY` interacts with the FEVM's memory management system. It is designed to handle the following scenarios:
+
+1. **Memory Expansion:** If the destination or source memory regions extend beyond the currently allocated memory, the memory is expanded as necessary to accommodate the operation.
+   
+2. **Overlapping Regions:** `MCOPY` ensures correct behavior when the source and destination regions overlap, copying data in a manner that avoids corruption.
+
+3. **Error Handling:** The opcode includes checks to prevent out-of-bounds memory access, raising appropriate errors if invalid memory regions are specified.
+
+## Design Rationale
```
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:51]:
```suggestion
## Design Rationale

```

File: FIPS/fip-mcopy-opcode-fevm.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# FIP-XXXX: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM
+
+## Simple Summary
+This proposal aims to introduce the `MCOPY` opcode, as defined in [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656), into the Filecoin Ethereum Virtual Machine (FEVM). This opcode allows for efficient memory copying within the FEVM, enhancing performance for smart contracts.
+
+## Abstract
+The `MCOPY` opcode provides a more efficient mechanism for memory copying within smart contracts. This proposal suggests integrating `MCOPY` into the FEVM, ensuring compatibility with Solidity versions v0.8.25 and later, which rely on this opcode for certain operations.
+
+## Change Motivation
+As Solidity continues to evolve, the use of the `MCOPY` opcode becomes increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.
```
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:23]:
```suggestion
## Simple Summary

This proposal aims to introduce the `MCOPY` opcode, as defined in [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656), into the Filecoin Ethereum Virtual Machine (FEVM). This opcode allows for efficient memory copying within the FEVM, enhancing performance for smart contracts.

## Abstract

The `MCOPY` opcode provides a more efficient mechanism for memory copying within smart contracts. This proposal suggests integrating `MCOPY` into the FEVM, ensuring compatibility with Solidity versions v0.8.25 and later, which rely on this opcode for certain operations.

## Change Motivation

As Solidity continues to evolve, the use of the `MCOPY` opcode becomes increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.
```

File: FIPS/fip-mcopy-opcode-fevm.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+## Backwards Compatibility
+The introduction of the `MCOPY` opcode does not affect existing contracts that do not utilize this opcode. Contracts compiled with Solidity versions prior to v0.8.25 will continue to function as intended without modification. However, contracts using `MCOPY` will require the opcode to be present in the FEVM to execute correctly.
+
+## Test Cases
+A series of test cases will be provided to ensure that the `MCOPY` opcode functions as expected under various scenarios. These test cases will include non-overlapping memory copy, overlapping memory copy, out-of-bounds source handling, and memory expansion.
+
+## Security Considerations
+Security implications for the `MCOPY` opcode are similar to other memory manipulation operations. Special care will be taken to ensure that overlapping memory regions are handled safely, preventing data corruption. Additionally, the implementation will include checks to prevent out-of-bounds memory access, which could otherwise lead to unexpected behavior or security vulnerabilities.
```
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:61]: >  Special care will be taken to ensure that overlapping memory regions are handled safely, preventing data corruption.

Elaborate on "the special care" that is taken to ensure the proper handling of memory region overlap

File: FIPS/fip-mcopy-opcode-fevm.md (hunk 3 above)
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:23]:
```suggestion
As Solidity continues to evolve, the use of the `MCOPY` opcode has become increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.
```

**Comment by jennijuju** [comment]
@snissn thanks for the FIP! This draft proposes a straightforward fevm enhancement and is well written!

**Review by jennijuju — APPROVED** [review]

File: FIPS/fip-mcopy-opcode-fevm.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+---
+fip: "<to be assigned>"
```
Comment (jennijuju) [inline FIPS/fip-mcopy-opcode-fevm.md:2]: please take 0094

**Review by TippyFlitsUK — APPROVED** [review]

File: FIPS/fip-mcopy-opcode-fevm.md (hunk 4 above)
Comment (snissn) [reply FIPS/fip-mcopy-opcode-fevm.md]: Sure - the the spec is clear that if we copy one region to another and they overlap it should work as if you take a copy of the source region and paste it over the destination region without any errors or bugs due to over-writes as if you had created a unique copy of the memory being copied before the write. The original EIP describes it as: `Copying takes place as if an intermediate buffer was used, allowing the destination and source to overlap.` in the built in actors implementation we use the slice built in method [copy_within](https://doc.rust-lang.org/stable/core/primitive.slice.html#method.copy_within) which accounts for that issue.

File: FIPS/fip-mcopy-opcode-fevm.md (hunk 1 above)
Comment (snissn) [reply FIPS/fip-mcopy-opcode-fevm.md]: Thanks! I was misunderstanding the meaning. I have removed that here: 0d7471f

**Comment by snissn** [comment]
Thanks @jennijuju I was able to update the PR with your suggestions. Please review the response that I hope resolves your question about `overlapping memory regions`.

File: README.md [hunk 6]
```diff
@@ -129,3 +129,4 @@ This improvement protocol helps achieve that objective for all members of the Fi
 | [0090](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0090.md) | Non-Interactive PoRep | FIP | luca (@lucaniz), kuba (@Kubuxu), nicola (@nicola), nemo (@cryptonemo), volker (@vmx), irene (@irenegia) | Superseded by [FIP0092](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0092.md) |
 | [0091](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0091.md) | Add support for Homestead and EIP-155 Ethereum Transactions ("legacy" Ethereum Transactions) | FIP | Aarsh (@aarshkshah1992)| Accepted |
 | [0092](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0092.md) | Non-Interactive PoRep | FIP | luca (@lucaniz), kuba (@Kubuxu), nicola (@nicola), nemo (@cryptonemo), volker (@vmx), irene (@irenegia), Alex North (@anorth), orjan (@Phi-rjan) | Final |
+| [XXXX](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-mcopy-opcode-fevm.md) | Add Support for EIP-5656 (MCOPY Opcode) in the FEVM | FIP | Michael Seiler (@snissn), Raúl Kripalani (@raulk), Steven Allen (@stebalien) | Draft |
```
Comment (jsoares) [inline README.md:132]:
```suggestion
| [0094](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-mcopy-opcode-fevm.md) | Add Support for EIP-5656 (MCOPY Opcode) in the FEVM | FIP | Michael Seiler (@snissn), Raúl Kripalani (@raulk), Steven Allen (@stebalien) | Draft |
```

**Review by jsoares — APPROVED** [review]

**Pull request merged**